src/transaction.rs

```diff
@@ -176,54 +103,94 @@ impl<D: Deliverable> Transaction<D> {
             requires_body,
         } = self;
-        trace!("Spawning request: {:?}", request);
+
+        let deliverable_guard = DeliverableDropGuard::new(deliverable);
 
         let start_time = Instant::now();
-        let task = task::current();
-        let request_future = client.request(request).and_then(move |response| {
-            if requires_body {
-                let (parts, body) = response.into_parts();
-                Either::A(
-                    body.fold(Vec::new(), |mut acc, chunk| {
-                        acc.extend_from_slice(&*chunk);
-                        future::ok::<_, hyper::Error>(acc)
-                    }).map(move |body| {
-                        let body_size = body.len();
-                        (
+
+        let request_future = async move {
+            trace!("Sending request: {:?}", request);
+            match client.request(request).await {
+                Ok(response) => {
+                    if requires_body {
+                        let (parts, mut body) = response.into_parts();
+                        let mut body_vec = Vec::new();
+
+                        while let Some(Ok(chunk)) = body.next().await {
+                            body_vec.extend_from_slice(&*chunk);
+                        }
+
+                        let body_size = body_vec.len();
+                        Ok((
                             Response::from_parts(parts, Body::empty()),
-                            Some(body),
+                            Some(body_vec),
                             body_size,
-                        )
-                    }),
-                )
-            } else {
-                // Note that you must consume the body if you want keepalive
-                // to take affect.
-                let (parts, body) = response.into_parts();
-                Either::B(
-                    body.fold(0, |acc, chunk| {
-                        future::ok::<_, hyper::Error>(acc + chunk.len())
-                    }).map(|body_size| {
-                        (Response::from_parts(parts, Body::empty()), None, body_size)
-                    }),
-                )
+                        ))
+                    } else {
+                        // Note that you must consume the body if you want keepalive
+                        // to take affect.
+                        let (parts, mut body) = response.into_parts();
+
+                        let mut body_len = 0;
+
+                        while let Some(Ok(chunk)) = body.next().await {
+                            body_len += chunk.len();
+                        }
+
+                        Ok((Response::from_parts(parts, Body::empty()), None, body_len))
+                    }
+                }
+                Err(e) => Err(e),
             }
-        });
+        };
 
-        let deadline = Instant::now() + timeout;
-        let work = Deadline::new(request_future, deadline);
+        let request_future = async move {
```

Comment:
```suggestion
        let result_future = async move {
```

Or some other name, just because shadowing the name `request_future` here tripped me up when reading since we already made a request in the future

Reply:
I'll just move the whole thing into the tokio::spawn